zooniverse/front-end-monorepo
# [WIP] Add service worker for classification POST

Package: lib-classifier

Closes #155

Describe your changes:
This adds a background sync service worker to handle failed classification POSTs 🎉 
WIP because 
- [ ] I need to add the config to the webpack production config.
- [ ] I need to check to see if the failure is a 422 in which case we will not add it to IndexDB and instead maybe report it to a log service?

Additionally, if for some reason the loading of the service worker and workbox, which is using the default config in the plugin to load over CDN, should we have the current PFE queue class as a fallback in this worst case scenario? We know it works and will provide queuing and error management in a rare case. 

There are several things to note:
- This is using Google's [Workbox for Webpack](https://developers.google.com/web/tools/workbox/guides/codelabs/webpack). Workbox is a set of predefined worker recipes to take out the boilerplate and reduce common gotchas. 
  - The specific configuration right now is using the `InjectManifest` plugin because this is the slightly more flexible configuration since it allows us to define our own worker, but it still handles the build steps to make sure the worker ends up in the root of the app. The `GenerateSW` plugin assumes you want a worker that caches assets, which we might want someday, but this was not the case today.
  - You can tell the `register` function to register service workers from other than the root, but I'm following the basic setup guidelines today. We can modify this another time. 
- The worker is registered in the `Classifier.js` component independently from the constructor of the component. This is preferred since eventually we may want to cache assets and we don't want the worker registration tied to a React component. 
- The Queue worker is in a new folder called `workers` as well as the registration function

How to test:

The [workbox-background-sync](https://developers.google.com/web/tools/workbox/modules/workbox-background-sync) instructions note under 'Testing Workbox Background Sync':
> Sadly, testing BackgroundSync sync is somewhat unintuitive and difficult for a number of reasons.

And provide a set of instructions to manually test the worker. I roughly followed that myself except as I coded the worker, I had to manually hit the update on the worker/clear cache. PR reviewers should be able to follow the instructions as described.

# Review Checklist

## General

- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
